Kudu.Core/Functions/FunctionManager.cs

```diff
@@ -49,52 +52,291 @@ public FunctionManager(IEnvironment environment, ITraceFactory traceFactory)
                         }
                     }
                 }
                 catch (Exception ex)
                 {
-                    tracer.Trace(String.Format("{0} is invalid. {1}", functionJson, ex.Message));
+                    tracer.Trace(String.Format("{0} is invalid. {1}", functionJson.Name, ex.Message));
                 }
             }
 
             return inputs;
         }
 
-        public IEnumerable<string> EnumerateFunctionFiles()
+        public async Task<FunctionEnvelope> CreateOrUpdate(string name, FunctionEnvelope functionEnvelope)
+        {
+            var functionDir = Path.Combine(_environment.FunctionsPath, name);
+
+            // Assert templateId and value are both present
+            if (!string.IsNullOrEmpty(functionEnvelope?.TemplateId))
+            {
+                // check to create template.
+                if (FileSystemHelpers.DirectoryExists(functionDir) &&
+                    FileSystemHelpers.GetFileSystemEntries(functionDir).Any())
+                {
+                    throw new InvalidOperationException($"Function {name} already exist");
+                }
+
+                var template = (await GetTemplatesFromGithub()).FirstOrDefault(e => e.name.Equals(functionEnvelope.TemplateId, StringComparison.OrdinalIgnoreCase));
+
+                if (template != null)
+                {
+                    await DeployTemplateFromGithub(template, functionDir);
+                }
+                else
+                {
+                    throw new FileNotFoundException($"template: {functionEnvelope.TemplateId} was not found");
+                }
+            }
+            else
+            {
+                // Create or update function
+                if (!FileSystemHelpers.DirectoryExists(functionDir))
+                {
+                    // create a new function
+                    FileSystemHelpers.EnsureDirectory(functionDir);
+                }
+                await FileSystemHelpers.WriteAllTextToFileAsync(Path.Combine(functionDir, Constants.FunctionsConfigFile), JsonConvert.SerializeObject(functionEnvelope?.Config ?? new JObject()));
+            }
+            return await GetFunctionConfig(name);
+        }
+
+        public async Task<IEnumerable<FunctionEnvelope>> ListFunctionsConfig()
+        {
+            var configList = await Task.WhenAll(
+                    FileSystemHelpers
+                    .GetDirectories(_environment.FunctionsPath)
+                    .Select(d => Path.Combine(d, Constants.FunctionsConfigFile))
+                    .Where(FileSystemHelpers.FileExists)
+                    .Select(async f => { try { return await GetFunctionConfig(Path.GetFileName(Path.GetDirectoryName(f))); } catch { return null; } }));
+            return configList.Where(c => c != null);
+        }
+
+        public async Task<FunctionEnvelope> GetFunctionConfig(string name)
```

Comment:
Maybe add a TryGetFunctionConfig overload that does the error / exception handling so you don't have to inline that above?